### PR TITLE
Extract values on the fly

### PR DESCRIPTION
**Pull request by ZacLN**
Here's the pseudo val-less branch I've been using, only extracting strings for identifiers and literals.

Some timings:
```
using Tokenize, BenchmarkTools
function speed_test()
    cnt = 0
    dir = dirname(Base.find_source_file("base.jl"))
    for (root, dirs, files) in walkdir(dir)
        for file in files
            if endswith(file, ".jl")
                file = joinpath(root, file)
                # println(file)
                str = readstring(file)
                l = tokenize(str)
                while !Tokenize.Lexers.eof(l)
                    t = Tokenize.Lexers.next_token(l)
                    if t.kind == Tokens.IDENTIFIER
                        cnt += 1
                    end
                end
            end
        end
    end
    cnt
end
@benchmark speed_test()
```
Before:
```
BenchmarkTools.Trial:
  memory estimate:  254.06 MiB
  allocs estimate:  5078640
  --------------
  minimum time:     615.183 ms (5.96% GC)
  median time:      618.250 ms (6.50% GC)
  mean time:        620.669 ms (6.29% GC)
  maximum time:     637.570 ms (6.46% GC)
  --------------
  samples:          9
  evals/sample:     1
```

After:
```
BenchmarkTools.Trial:
  memory estimate:  154.10 MiB
  allocs estimate:  2633640
  --------------
  minimum time:     333.714 ms (6.79% GC)
  median time:      338.120 ms (6.88% GC)
  mean time:        337.438 ms (6.87% GC)
  maximum time:     340.331 ms (6.85% GC)
  --------------
  samples:          15
  evals/sample:     1
```

and obviously it disables the round-tripping tests

### CODE REVIEW

**Comment by KristofferC** [comment]
Nice. This is a little bit too tailor made for the parser though? I would prefer to have two modes:

1) Completely roundtrippable (but still as efficient as possible). This can be done similar to how it is done here, except also store `val` for whitespace and comments.

2) No vals at all. Tokens can then be stack allocated, inlined into vectors and there is no need to write to any buffer. If you need a `val` later, you can extract it from the original string by using the byte offsets in each Token.

Opinions? cc @keno

**Comment by ZacLN** [comment]
Yep that sounds fine w/ me, we skip ws/comments internally in CSTParser anyway. Do we need to add  `RawToken <: Token`  and `ValToken <: Token`for the second case?

**Comment by KristofferC** [comment]
For backwards compatability, I think we should keep `Token` as is, add an `AbstractToken` and a `RawToken`.

File: src/lexer.jl [hunk 1]
```diff
@@ -215,15 +216,23 @@ end
...
+    elseif kind == Tokens.ERROR
+        str = String(l.io.data[(l.token_startpos + 1):position(l.io)])
     else
         str = ""
     end
-    tok = Token(kind, (l.token_start_row, l.token_start_col),
-                (l.current_row, l.current_col - 1),
-                startpos(l), position(l) - 1,
-                str, err)
+    if l.raw
```
Comment (KristofferC) [inline src/lexer.jl]: This will not be type stable right.

Comment (ZacLN) [reply src/lexer.jl]: Yep return type is a Union, I can just parameterize `Lexer` if you like

**Comment by Keno** [comment]
> I can just parameterize Lexer if you like

I think that would be the best way to go about it.

File: src/lexer.jl [hunk 2]
```diff
@@ -34,24 +34,25 @@ mutable struct Lexer{IO_t <: IO}
     charstore::IOBuffer
     current_char::Char
     doread::Bool
-    raw::Bool
 end
 
-Lexer(io, raw) = Lexer(io, position(io), 1, 1, position(io), 1, 1, position(io), Tokens.ERROR, IOBuffer(), ' ', false, raw)
+Lexer(io::IO_t, raw) where {IO_t} = Lexer{IO_t,raw ? RawToken : Token}(io, position(io), 1, 1, position(io), 1, 1, position(io), Tokens.ERROR, IOBuffer(), ' ', false)
```
Comment (KristofferC) [inline src/lexer.jl]: I think the second argument to the `Lexer` / `tokenize` has to be a type in order for everything to be type stable (including the creation of the Lexer). So the constructor would be e.g. `Lexer("foo", Token)` or `Lexer("foo", RawToken)`.

Comment (ZacLN) [reply src/lexer.jl]: Doesn't appear to make a difference, but I'll change it. 
Edit: do you mean funcs. w/ signatures: `tokenize(str::String,::Type{Token})` ?

Comment (KristofferC) [reply src/lexer.jl]: Yes, I would assume the constant folding would work if you have `::Union{Type{Token}, Type{RawToken}}` and have a conditional on the type but perhaps you need two function definitions.

**Comment by ZacLN** [comment]
All sorted, anything else?

File: test/lexer.jl [hunk 3]
```diff
@@ -34,7 +30,7 @@ end # testset
         token_strs = ["𝘋", " ", "=", "2", "β", ""]
         for (i, n) in enumerate(l)
             @test T.kind(n) == kinds[i]
-            @test untokenize(n)  == token_strs[i]
+            # @test untokenize(n)  == token_strs[i]
```
Comment (KristofferC) [inline test/lexer.jl]: This doesn't have to be removed now?

File: src/lexer.jl [hunk 4]
```diff
@@ -34,23 +32,29 @@ mutable struct Lexer{IO_t <: IO}
...
 end
 
-Lexer(io) = Lexer(io, position(io), 1, 1, -1, position(io), 1, 1, position(io), Tokens.ERROR, IOBuffer())
-Lexer(str::AbstractString) = Lexer(IOBuffer(str))
+Lexer(io::IO_t, T) where {IO_t} = Lexer{IO_t,T}(io, position(io), 1, 1, position(io), 1, 1, position(io), Tokens.ERROR, IOBuffer(), ' ', false)
+Lexer(str::AbstractString, T) = Lexer(IOBuffer(str), T)
 
 """
-    tokenize(x)
+    tokenize(x, raw::Bool = false)
```
Comment (KristofferC) [inline src/lexer.jl]: This docstring is not accurate anymore?

**Review by KristofferC — COMMENTED** [review]
Perhaps add a few very basic tests for the RawToken Lexer (just so the code paths are hit) and perhaps an `@inferred` for `untokenize` with `RawToken`?

**Comment by ZacLN** [comment]
Ta, fixes made

**Pull request closed without merging**